cli/sql.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Marc berhault (marc@cockroachlabs.com)
+
+package cli
+
+import (
+	"database/sql"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"text/tabwriter"
+
+	// Import cockroach driver.
+	_ "github.com/cockroachdb/cockroach/sql/driver"
+	"github.com/spf13/cobra"
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+func makeSQLClient() *sql.DB {
+	// TODO(pmattis): Initialize the user to something more
+	// reasonable. Perhaps Context.Addr should be considered a URL.
+	db, err := sql.Open("cockroach",
+		fmt.Sprintf("%s://root@%s?certs=%s",
+			Context.RequestScheme(),
+			Context.Addr,
+			Context.Certs))
+	if err != nil {
+		fmt.Fprintf(osStderr, "failed to initialize SQL client: %s\n", err)
+		osExit(1)
+	}
+	return db
+}
+
+// sqlShellCmd opens a sql shell.
+var sqlShellCmd = &cobra.Command{
+	Use:   "sql [options]",
+	Short: "open a sql shell",
+	Long: `
+Open a sql shell running against the cockroach database at --addr.
+`,
+	Run: runTerm,
+}
+
+var isQuery = map[string]struct{}{
+	"SELECT": {},
+	"SHOW":   {},
+}
+
+func runTerm(cmd *cobra.Command, args []string) {
+	if len(args) != 0 {
+		cmd.Usage()
+		return
+	}
+
+	db := makeSQLClient()
+
+	readWriter := struct {
+		io.Reader
+		io.Writer
+	}{
+		Reader: os.Stdin,
+		Writer: os.Stdout,
+	}
+
+	oldState, err := terminal.MakeRaw(0)
+	if err != nil {
+		panic(err)
+	}
+	defer func() {
+		_ = terminal.Restore(0, oldState)
+	}()
+
+	term := terminal.NewTerminal(readWriter, "> ")
+	for {
+		line, err := term.ReadLine()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Err: %s\n", err)
+			osExit(1)
+		}
+		words := strings.Split(line, " ")
+		if len(words) == 0 {
+			continue
+		}
+
+		// This is a really terrible way of figuring out how to handle
+		// the input.
+		command := strings.ToUpper(words[0])
```

Comment:
I'm not suggesting this needs to be done in this PR, but that it should be the roadmap. Yes, we can't currently specify the database in the URL, but we should fix that. Both the mysql and postgres golang drivers allow specifying the database that way (albeit via different formats).